File: CPP/ByRound/1985/1985_H1_Maximize_the_Largest_Component_Easy_Version.cpp
// Problem: CF 1985 H1 - Maximize the Largest Component (Easy Version)
// https://codeforces.com/contest/1985/problem/H1

/*
 * Problem: Maximize the Largest Component (Easy Version)
 * Algorithm: 
 *   - Preprocess the grid to count empty cells in each row and column
 *   - For each connected component of '#' cells, simulate the effect of setting
 *     an entire row or column to '#' and calculate how much it increases the
 *     size of the largest connected component.
 *   - Use BFS to find connected components and update the contribution of each
 *     row/column to the final answer.
 *
 * Time Complexity: O(n * m) per test case
 * Space Complexity: O(n * m) for the grid and auxiliary arrays
 */

#include <stddef.h>
#include <algorithm>
#include <iostream>
#include <queue>
#include <utility>

using namespace std;
void initial() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
}
using ll = long long int;
void run() {
  int n, m, i, j, mau, mad, mar, mal, f;
  cin >> n >> m;
  bool a[n + 2][m + 2]; // Grid with padding
  int b[n + 1], c[m + 1]; // b[i] = number of empty cells in row i, c[j] = number of empty cells in column j
  char t;
  // Read the grid and set padding to 0
  for (i = 1; i <= n; i++) {
    for (j = 1; j <= m; j++)
      cin >> t, a[i][j] = (t == '#');
  }
  for (i = 0; i <= n + 1; i++)
    a[i][0] = 0, a[i][m + 1] = 0;
  for (i = 0; i <= m + 1; i++)
    a[0][i] = 0, a[n + 1][i] = 0;
  // Initialize row and column counts
  for (i = 1; i <= n; i++)
    b[i] = 0;
  for (i = 1; i <= m; i++)
    c[i] = 0;
  // Count empty cells in each row
  for (i = 1; i <= n; i++) {
    for (j = 1; j <= m; j++)
      b[i] += !a[i][j];
  }
  // Count empty cells in each column
  for (i = 1; i <= m; i++) {
    for (j = 1; j <= n; j++)
      c[i] += !a[j][i];
  }
  queue<pair<int, int>> q; // BFS queue
  // Process each connected component of '#' cells
  for (i = 1; i <= n; i++) {
    for (j = 1; j <= m; j++) {
      if (!a[i][j]) // Skip if already processed or is '.' 
        continue;
      f = 0; // Size of the connected component
      mau = i; // Min row
      mad = i; // Max row
      mal = j; // Min col
      mar = j; // Max col
      // BFS to find the connected component
      for (q.push(make_pair(i, j)); !q.empty(); q.pop()) {
        if (!a[q.front().first][q.front().second]) // Skip if already processed  
          continue;
        mau = min(mau, q.front().first);
        mad = max(mad, q.front().first);
        mal = min(mal, q.front().second);
        mar = max(mar, q.front().second);
        a[q.front().first][q.front().second] = 0; // Mark as processed
        f++; // Increment size
        // Push neighbors to queue if they are part of the component
        if (a[q.front().first][q.front().second + 1])
          q.push(make_pair(q.front().first, q.front().second + 1));
        if (a[q.front().first][q.front().second - 1])
          q.push(make_pair(q.front().first, q.front().second - 1));
        if (a[q.front().first + 1][q.front().second])
          q.push(make_pair(q.front().first + 1, q.front().second));
        if (a[q.front().first - 1][q.front().second])
          q.push(make_pair(q.front().first - 1, q.front().second));
      }
      // Increment the row and column counts with the size of the component
      mau--; // Adjust for 1-based indexing
      mal--;
      mad++;
      mar++;
      while (mau <= mad)
        b[mau] += f, mau++;
      while (mal <= mar)
        c[mal] += f, mal++;
    }
  }
  // Find the maximum count in rows and columns
  mau = 0;
  for (i = 1; i <= n; i++)
    mau = max(mau, b[i]);
  for (i = 1; i <= m; i++)
    mau = max(mau, c[i]);
  cout << mau << '\n';
}
int main() {
  initial();
  int tt, i, n;
  cin >> tt;
  while (tt--)
    run();
}


// https://github.com/VaHiX/CodeForces/

File: CSharp/ByRound/1985/1985_H1_Maximize_the_Largest_Component_Easy_Version.cs
// Problem: CF 1985 H1 - Maximize the Largest Component (Easy Version)
// https://codeforces.com/contest/1985/problem/H1

/*
 * Algorithm: Connected Components and Grid Operations
 * 
 * Purpose: Find the maximum size of a connected component of '#' cells after performing at most one operation of setting an entire row or column to '#'.
 * 
 * Approach:
 * 1. First, identify all connected components of '#' cells using BFS and label them.
 * 2. For each row, simulate setting all cells in that row to '#' and calculate the potential maximum component size:
 *    - Count how many distinct components are adjacent to the '.' cells in that row.
 *    - Sum the sizes of these components with the area of '.' cells in the row.
 * 3. Do the same for each column.
 * 4. Return the maximum among all possibilities.
 * 
 * Time Complexity: O(n * m) per test case, where n is rows and m is columns.
 * Space Complexity: O(n * m) for storing grid, components labels, and component sizes.
 */

using System;
using System.Collections.Generic;
using System.IO;
class Program {
  static (int, int)[] Directions = { (-1, 0), (0, 1), (1, 0), (0, -1) };
  static bool IsInBounds((int, int)point, int rowCount, int columnCount) {
    return point.Item1 >= 0 && point.Item1 < rowCount && point.Item2 >= 0 &&
           point.Item2 < columnCount;
  }
  static void FillGrid(List<string> grid, int rowCount, StreamReader reader) {
    for (int i = 0; i < rowCount; i++) {
      grid.Add(reader.ReadLine());
    }
  }
  static void SolveTestCase(int testCaseIndex, StreamReader reader,
                            StreamWriter writer) {
    int rowCount, columnCount;
    string[] dimensions = reader.ReadLine().Split();
    rowCount = int.Parse(dimensions[0]);
    columnCount = int.Parse(dimensions[1]);
    List<string> grid = new List<string>();
    FillGrid(grid, rowCount, reader);
    long[,] components = new long[rowCount, columnCount];
    Dictionary<long, long> componentSizes = new Dictionary<long, long>();
    long maxComponentSize = 0, componentLabel = 1;
    // First pass: Label all connected components of '#' cells
    for (int i = 0; i < rowCount; i++) {
      for (int j = 0; j < columnCount; j++) {
        if (grid[i][j] == '#' && components[i, j] == 0) {
          Queue<(int, int)> queue = new Queue<(int, int)>();
          queue.Enqueue((i, j));
          while (queue.Count > 0) {
            var currentPoint = queue.Dequeue();
            if (!IsInBounds(currentPoint, rowCount, columnCount)) {
              continue;
            }
            if (grid[currentPoint.Item1][currentPoint.Item2] == '.' ||
                components[currentPoint.Item1, currentPoint.Item2] != 0) {
              continue;
            }
            components[currentPoint.Item1, currentPoint.Item2] = componentLabel;
            if (!componentSizes.ContainsKey(componentLabel))
              componentSizes[componentLabel] = 1;
            else
              componentSizes[componentLabel]++;
            foreach (var direction in Directions) {
              queue.Enqueue((currentPoint.Item1 + direction.Item1,
                             currentPoint.Item2 + direction.Item2));
            }
          }
          maxComponentSize =
              Math.Max(maxComponentSize, componentSizes[componentLabel]);
          componentLabel++;
        }
      }
    }
    // Second pass: Evaluate maximum component size if we fill a row
    for (int i = 0; i < rowCount; i++) {
      long totalArea = 0;
      HashSet<long> componentLabels = new HashSet<long>();
      for (int j = 0; j < columnCount; j++) {
        if (grid[i][j] == '#') {
          componentLabels.Add(components[i, j]);
        } else {
          totalArea++; // Count the '.' cell in the row
          // Check adjacent '#' cells (up and down)
          if (i > 0 && grid[i - 1][j] == '#') {
            componentLabels.Add(components[i - 1, j]);
          }
          if (i + 1 < rowCount && grid[i + 1][j] == '#') {
            componentLabels.Add(components[i + 1, j]);
          }
        }
      }
      foreach (long label in componentLabels) {
        totalArea += componentSizes[label];
      }
      maxComponentSize = Math.Max(maxComponentSize, totalArea);
    }
    // Third pass: Evaluate maximum component size if we fill a column
    for (int j = 0; j < columnCount; j++) {
      long totalArea = 0;
      HashSet<long> componentLabels = new HashSet<long>();
      for (int i = 0; i < rowCount; i++) {
        if (grid[i][j] == '#') {
          componentLabels.Add(components[i, j]);
        } else {
          totalArea++; // Count the '.' cell in the column
          // Check adjacent '#' cells (left and right)
          if (j > 0 && grid[i][j - 1] == '#') {
            componentLabels.Add(components[i, j - 1]);
          }
          if (j + 1 < columnCount && grid[i][j + 1] == '#') {
            componentLabels.Add(components[i, j + 1]);
          }
        }
      }
      foreach (long label in componentLabels) {
        totalArea += componentSizes[label];
      }
      maxComponentSize = Math.Max(maxComponentSize, totalArea);
    }
    writer.WriteLine(maxComponentSize);
  }
  static void Main(string[] args) {
    using (
        StreamReader reader = new StreamReader(
            Console
                .OpenStandardInput())) using (StreamWriter writer =
                                                  new StreamWriter(
                                                      Console
                                                          .OpenStandardOutput())) {
      int testCaseCount = int.Parse(reader.ReadLine());
      for (int i = 0; i < testCaseCount; i++) {
        SolveTestCase(i, reader, writer);
      }
    }
  }
}


// https://github.com/VaHiX/CodeForces/